hibernate-core/src/main/java/org/hibernate/sql/Template.java

```diff
@@ -205,6 +206,34 @@ else if ( quotedIdentifier && ( dialect.closeQuote()==token.charAt(0) ) ) {
 				continue;
 			}
 
+			// Special processing for type casting
+			if ( inCast && "as".equals( lcToken ) ) {
+				result.append( " as " );
+
+				hasMore = tokens.hasMoreTokens();
+				nextToken = hasMore ? tokens.nextToken() : null;
+
```

Comment:
According to SQL92, the target type itself can contain parentheses. Examples are: CHARACTER(10), NUMERIC(10,4). IIUC, inCast will be set to false when the first right-parentheses is found. The code should detect the right-parentheses following the data type (which may contain parentheses) before setting inCast to false. 

It looks like the simple case with a target data type that does not have parentheses is covered.